lib/h3core.js

```diff
@@ -253,57 +326,70 @@ function destroyGeoPolygon(geoPolygon) {
- * intended to be the invocation that caused the side effect, e.g. readLong(H3.getSomeLong())
+ * Read an H3 index from a pointer to C memory.
  * @private
- * @param  {number} invocation Invoked function returning a long value. The actual return
- *                             value of these functions is a 32-bit integer.
- * @return {number}            Long value as a [lower, upper] pair
+ * @param  {number} cAddress  Pointer to allocated C memory
+ * @param {number} offset     Offset, in number of H3 indexes, in case we're
+ *                            reading an array
+ * @return {H3Index}          H3 index, or null if index was invalid
  */
-function readLong(invocation) {
-    // Upper 32-bits of the long set via side-effect
-    const upper = C.getTempRet0();
-    return [invocation, upper];
+function readH3IndexFromPointer(cAddress, offset = 0) {
+    const lower = C.getValue(cAddress + SZ_INT * offset * 2, 'i32');
+    const upper = C.getValue(cAddress + SZ_INT * (offset * 2 + 1), 'i32');
+    // The lower bits are allowed to be 0s, but if the upper bits are 0
+    // this represents an invalid H3 index
+    return upper ? splitLongToH3Index(lower, upper) : null;
 }
 
 /**
- * Read an H3 index from a C return value. As with readLong, the argument to this function
- * is intended to be an invocation, e.g. readH3Index(H3.getSomeAddress()), to help ensure that
- * the temp value storing the upper bits of the long is still set.
+ * Read a boolean from a pointer to C memory.
  * @private
- * @param  {number} invocation  Invoked function returning a single H3 index
- * @return {H3Index}            H3 index, or null if index was invalid
+ * @param  {number} cAddress  Pointer to allocated C memory
+ * @param {number} offset     Offset, in number of booleans, in case we're
+ *                            reading an array
+ * @return {Boolean} Boolean value
  */
-function readH3Index(invocation) {
-    const [lower, upper] = readLong(invocation);
-    // The lower bits are allowed to be 0s, but if the upper bits are 0
-    // this represents an invalid H3 index
-    return upper ? splitLongToh3Index(lower, upper) : null;
+function readBooleanFromPointer(cAddress, offset = 0) {
+    const val = C.getValue(cAddress + SZ_INT * offset * 2, 'i32');
+    return Boolean(val);
 }
 
 /**
- * Read an H3 index from a pointer to C memory.
+ * Read a double from a pointer to C memory.
  * @private
  * @param  {number} cAddress  Pointer to allocated C memory
- * @param {number} offset     Offset, in number of H3 indexes, in case we're
+ * @param {number} offset     Offset, in number of booleans, in case we're
  *                            reading an array
- * @return {H3Index}          H3 index, or null if index was invalid
+ * @return {number} Double value
  */
-function readH3IndexFromPointer(cAddress, offset = 0) {
+function readDoubleFromPointer(cAddress, offset = 0) {
+    return C.getValue(cAddress + SZ_DBL * offset * 2, 'double');
```

Comment:
The number of cells will not be returned as a double so this will not work for that. You could perhaps pass in i64 (https://emscripten.org/docs/api_reference/preamble.js.html?highlight=getvalue#getValue) which might (I haven't tested this) convert to a JS double.